From e46a898017d18b12a6faef2b4c17d0b0f2234774 Mon Sep 17 00:00:00 2001
From: Hsin-chen Chuang <chharry@google.com>
Date: Wed, 17 Sep 2025 17:45:32 +0800
Subject: [PATCH] UPSTREAM: Fix the format of AdapterBondedDevices

The format has been changed by http://ag/33020682. Theoretically when
there are more than 1 bonded devices then we would get the wrong
addresses starting from the 2nd.

Also fixed some clippy warnings.

Bug: 432412937
Test: m com.google.android.bt
Test: floss/build/build-in-container.py
Test: bluetooth_AdapterSRHealth.sr_reconnect_multiple_classic_hid.floss
Test: bluetooth_AdapterQuickHealth.AVL.all_floss
Flag: EXEMPT BUGFIX
Change-Id: Ie84bfd4c98bf4806fc498ac42673d824f25fb6c6
---
 floss/rust/topshim/src/btif.rs | 130 ++++++++++++++++++++-------------
 1 file changed, 78 insertions(+), 52 deletions(-)

diff --git a/floss/rust/topshim/src/btif.rs b/floss/rust/topshim/src/btif.rs
index a0d9ae7511d..eb90a2b3946 100644
--- a/floss/rust/topshim/src/btif.rs
+++ b/floss/rust/topshim/src/btif.rs
@@ -59,5 +59,5 @@ impl From<u8> for BtTransport {
 
-impl Into<u8> for BtTransport {
-    fn into(self) -> u8 {
-        self.to_u8().unwrap_or(0)
+impl From<BtTransport> for u8 {
+    fn from(val: BtTransport) -> Self {
+        val.to_u8().unwrap_or(0)
     }
@@ -254,6 +254,6 @@ pub fn ascii_to_string(data: &[u8], length: usize) -> String {
         .take_while(|&(pos, &c)| c != 0 && pos < length)
-        .map(|(_pos, &x)| x.clone())
+        .map(|(_pos, &x)| x)
         .collect::<Vec<u8>>();
 
-    return String::from_utf8(ascii).unwrap_or_default();
+    String::from_utf8(ascii).unwrap_or_default()
 }
@@ -270,3 +270,3 @@ fn u16_from_bytes(item: &[u8]) -> u16 {
     let len = std::cmp::min(item.len(), 2);
-    u[0..len].copy_from_slice(&item);
+    u[0..len].copy_from_slice(item);
     u16::from_ne_bytes(u)
@@ -283,5 +283,5 @@ impl From<bindings::bt_status_t> for BtStatus {
 
-impl Into<u32> for BtStatus {
-    fn into(self) -> u32 {
-        self.to_u32().unwrap_or_default()
+impl From<BtStatus> for u32 {
+    fn from(val: BtStatus) -> Self {
+        val.to_u32().unwrap_or_default()
     }
@@ -289,5 +289,5 @@ impl Into<u32> for BtStatus {
 
-impl Into<i32> for BtStatus {
-    fn into(self) -> i32 {
-        self.to_i32().unwrap_or_default()
+impl From<BtStatus> for i32 {
+    fn from(val: BtStatus) -> Self {
+        val.to_i32().unwrap_or_default()
     }
@@ -310,3 +310,3 @@ impl From<bindings::bt_service_record_t> for BtServiceRecord {
     fn from(item: bindings::bt_service_record_t) -> Self {
-        let name = item.name.iter().map(|&x| x.clone() as u8).collect::<Vec<u8>>();
+        let name = item.name.iter().map(|&x| x as u8).collect::<Vec<u8>>();
 
@@ -335,5 +335,5 @@ impl From<bindings::bt_scan_mode_t> for BtScanMode {
 
-impl Into<bindings::bt_scan_mode_t> for BtScanMode {
-    fn into(self) -> bindings::bt_scan_mode_t {
-        BtScanMode::to_u32(&self).unwrap_or_default()
+impl From<BtScanMode> for bindings::bt_scan_mode_t {
+    fn from(val: BtScanMode) -> Self {
+        BtScanMode::to_u32(&val).unwrap_or_default()
     }
@@ -356,5 +356,5 @@ impl From<u32> for BtDiscMode {
 
-impl Into<u32> for BtDiscMode {
-    fn into(self) -> u32 {
-        self.to_u32().unwrap_or(0)
+impl From<BtDiscMode> for u32 {
+    fn from(val: BtDiscMode) -> Self {
+        val.to_u32().unwrap_or(0)
     }
@@ -410,5 +410,5 @@ impl From<u32> for BtAddrType {
 
-impl Into<u32> for BtAddrType {
-    fn into(self) -> u32 {
-        self.to_u32().unwrap_or(0)
+impl From<BtAddrType> for u32 {
+    fn from(val: BtAddrType) -> Self {
+        val.to_u32().unwrap_or(0)
     }
@@ -422,5 +422,5 @@ impl From<u8> for BtAddrType {
 
-impl Into<u8> for BtAddrType {
-    fn into(self) -> u8 {
-        self.to_u8().unwrap_or(0)
+impl From<BtAddrType> for u8 {
+    fn from(val: BtAddrType) -> Self {
+        val.to_u8().unwrap_or(0)
     }
@@ -434,5 +434,6 @@ pub type BtVendorProductInfo = bindings::bt_vendor_product_info_t;
 
-impl ToString for BtVendorProductInfo {
-    fn to_string(&self) -> String {
-        format!(
+impl Display for BtVendorProductInfo {
+    fn fmt(&self, f: &mut Formatter) -> Result {
+        write!(
+            f,
             "{}:v{:04X}p{:04X}d{:04X}",
@@ -454,3 +455,3 @@ impl TryFrom<Uuid> for Vec<u8> {
     fn try_from(value: Uuid) -> std::result::Result<Self, Self::Error> {
-        Ok((&value.uu).to_vec())
+        Ok(value.uu.to_vec())
     }
@@ -519,3 +520,3 @@ impl Uuid {
 
-        let raw = raw.chars().filter(|c| c.is_digit(16)).collect::<String>();
+        let raw = raw.chars().filter(|c| c.is_ascii_hexdigit()).collect::<String>();
         let s = raw.as_str();
@@ -576,3 +577,3 @@ impl Display for Uuid {
 pub struct DisplayUuid<'a>(pub &'a Uuid);
-impl<'a> Display for DisplayUuid<'a> {
+impl Display for DisplayUuid<'_> {
     fn fmt(&self, f: &mut Formatter) -> Result {
@@ -631,2 +632,7 @@ const PROPERTY_NAME_MAX: usize = 255;
 
+/// This is the length of tBLE_BD_ADDR_SERIALIZED, the prop format of AdapterBondedDevices.
+/// Instead of using mem::size_of::<bindings::tBLE_BD_ADDR_SERIALIZED>(), we hardcode this so that
+/// we can discover this easier (crash!) when the layout is changed.
+const TYPED_ADDR_LENGTH: usize = bindings::RawAddress_kLength as usize + 1;
+
 impl BluetoothProperty {
@@ -662,4 +668,5 @@ impl BluetoothProperty {
 
+    /// Returns the length when converted to bt_property_t
     fn get_len(&self) -> usize {
-        match &*self {
+        match self {
             BluetoothProperty::BdName(name) => cmp::min(PROPERTY_NAME_MAX, name.len() + 1),
@@ -672,5 +679,3 @@ impl BluetoothProperty {
             }
-            BluetoothProperty::AdapterBondedDevices(devlist) => {
-                devlist.len() * mem::size_of::<RawAddress>()
-            }
+            BluetoothProperty::AdapterBondedDevices(devlist) => devlist.len() * TYPED_ADDR_LENGTH,
             BluetoothProperty::AdapterDiscoverableTimeout(_) => mem::size_of::<u32>(),
@@ -702,3 +707,3 @@ impl BluetoothProperty {
         let len = self.get_len();
-        match &*self {
+        match self {
             BluetoothProperty::BdName(name) => {
@@ -741,5 +746,11 @@ impl BluetoothProperty {
                 for (idx, &dev) in devlist.iter().enumerate() {
-                    let start = idx * mem::size_of::<RawAddress>();
-                    let end = idx + mem::size_of::<RawAddress>();
+                    let start = idx * TYPED_ADDR_LENGTH;
+                    let end = start + mem::size_of::<RawAddress>();
                     data[start..end].copy_from_slice(&dev.address);
+                    // 0 is public address. We never pass this prop from Rust to LibBluetooth so
+                    // this shouldn't matter for now.
+                    log::error!(
+                        "Converting BluetoothProperty::AdapterBondedDevices to bt_property_t"
+                    );
+                    data[end] = 0;
                 }
@@ -839,7 +850,21 @@ impl From<bindings::bt_property_t> for BluetoothProperty {
             BtPropertyType::AdapterBondedDevices => {
-                let count = len / mem::size_of::<RawAddress>();
-                BluetoothProperty::AdapterBondedDevices(ptr_to_vec(
-                    prop.val as *const RawAddress,
-                    count,
-                ))
+                assert!(
+                    len % TYPED_ADDR_LENGTH == 0,
+                    "Invalid AdapterBondedDevices prop len: {}",
+                    len
+                );
+                let count = len / TYPED_ADDR_LENGTH;
+                BluetoothProperty::AdapterBondedDevices(
+                    (0..count)
+                        .map(|idx| {
+                            // The prop is an array of tBLE_BD_ADDR_SERIALIZED which has length
+                            // TYPED_ADDR_LENGTH. The first 6 bytes are the RawAddress and the 7th
+                            // is the address type (public / random / etc), while we don't care
+                            // about it for now.
+                            let start = idx * TYPED_ADDR_LENGTH;
+                            let end = start + mem::size_of::<RawAddress>();
+                            RawAddress::from_bytes(&slice[start..end]).unwrap_or_default()
+                        })
+                        .collect(),
+                )
             }
@@ -854,3 +879,3 @@ impl From<bindings::bt_property_t> for BluetoothProperty {
                 let v = unsafe { (prop.val as *const BtRemoteVersion).read_unaligned() };
-                BluetoothProperty::RemoteVersionInfo(v.clone())
+                BluetoothProperty::RemoteVersionInfo(v)
             }
@@ -858,3 +883,3 @@ impl From<bindings::bt_property_t> for BluetoothProperty {
                 let v = unsafe { (prop.val as *const BtLocalLeFeatures).read_unaligned() };
-                BluetoothProperty::LocalLeFeatures(v.clone())
+                BluetoothProperty::LocalLeFeatures(v)
             }
@@ -965,5 +990,6 @@ impl Hash for RawAddress {
 
-impl ToString for RawAddress {
-    fn to_string(&self) -> String {
-        format!(
+impl Display for RawAddress {
+    fn fmt(&self, f: &mut Formatter) -> Result {
+        write!(
+            f,
             "{:02X}:{:02X}:{:02X}:{:02X}:{:02X}:{:02X}",
@@ -987,3 +1013,3 @@ impl RawAddress {
         raw.copy_from_slice(raw_addr);
-        return Some(RawAddress { address: raw });
+        Some(RawAddress { address: raw })
     }
@@ -1012,3 +1038,3 @@ impl RawAddress {
     pub fn to_byte_arr(&self) -> [u8; 6] {
-        self.address.clone()
+        self.address
     }
@@ -1022,3 +1048,3 @@ impl RawAddress {
 pub struct DisplayAddress<'a>(pub &'a RawAddress);
-impl<'a> Display for DisplayAddress<'a> {
+impl Display for DisplayAddress<'_> {
     fn fmt(&self, f: &mut Formatter) -> Result {
@@ -1528,3 +1554,3 @@ mod tests {
         for (i, v) in slice.iter().enumerate() {
-            bdname.name[i] = v.clone();
+            bdname.name[i] = *v;
         }
